src/lib.rs

```diff
@@ -412,7 +412,7 @@ pub enum ScalarValue {
 }
 
 /// Additional information, dependendent on the kind of constant.
-#[derive(Debug, PartialEq)]
+#[derive(Debug, PartialEq, Clone)]
```

Comment:
why does this need a `Clone`?

Reply:
I made it `Clone` for convenience so that for example in `cast` we don't need to redefine a `ConstantInner` in each match arm

Reply:
where is that code that uses it?

Reply:
Lines 179 and 219.
For example `unary_op` has
```rust
ConstantInner::Composite {
    ty,
    ref mut components,
} => {
    match self.types[ty].inner {
        crate::TypeInner::Vector { .. } | crate::TypeInner::Matrix { .. } => (),
        _ => return Err(ConstantSolvingError::InvalidCastArg),
    }

    for component in components {
        *component = self.unary_op(op, *component)?
    }
}
```

Instead we would need
```rust
ConstantInner::Composite {
    ty,
    ref components,
} => {
    match self.types[ty].inner {
        crate::TypeInner::Vector { .. } | crate::TypeInner::Matrix { .. } => (),
        _ => return Err(ConstantSolvingError::InvalidCastArg),
    }

   let mut new_components = Vec::new();

    for component in components {
        let new_component = self.unary_op(op, *component)?
        new_components.push(new_component)
    }
    
    ConstantInner::Composite {
       ty,
       components: new_components
    }
}
```

Reply:
Yeah, I think it would be worth it. The code clearly becomes more functional, less imperative, less mutating, etc.
It's an improvement, and it's not that much more code to write.